File: src/main/java/org/radarcns/consumer/realtime/action/appserver/AppserverClient.java
package org.radarcns.consumer.realtime.action.appserver;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import java.io.IOException;
import java.net.MalformedURLException;
import java.net.URI;
import java.net.URL;
import java.util.Map;
import okhttp3.Headers;
import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import org.radarbase.exception.TokenException;
import org.radarbase.oauth.OAuth2Client;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class AppserverClient {

  private static final Logger logger = LoggerFactory.getLogger(AppserverClient.class);

  private final String baseUrl;
  private final OkHttpClient httpClient;
  private final ObjectMapper objectMapper;
  private final OAuth2Client oauthClient;

  public AppserverClient(String baseUrl, String tokenUrl, String clientId, String clientSecret)
      throws MalformedURLException {
    this.baseUrl = baseUrl;
    httpClient = new OkHttpClient.Builder().build();
    objectMapper = new ObjectMapper();

    if (tokenUrl == null || tokenUrl.isEmpty()) {
      logger.warn("MP Token url was not provided. Will use unauthenticated requests to appserver.");
      this.oauthClient = null;
    } else {
      if (clientId.isEmpty()) {
        throw new IllegalArgumentException("Client ID for appserver client is not set.");
      }
      this.oauthClient =
          new OAuth2Client.Builder()
              .credentials(clientId, clientSecret)
              .endpoint(new URL(tokenUrl))
              .scopes("SUBJECT.READ SUBJECT.UPDATE PROJECT.READ")
              .httpClient(httpClient)
              .build();
    }
  }

  protected Headers headers() throws TokenException {
    if (oauthClient != null) {
      return Headers.of("Authorization", "Bearer " + oauthClient.getValidToken().getAccessToken());
    } else {
      return Headers.of();
    }
  }

  public Map<String, Object> createMessage(
      String projectId, String userId, String type, String body) throws IOException {
    URI uri =
        URI.create(baseUrl)
            .resolve("/projects/" + projectId + "/users/" + userId + "/messaging/" + type);

    RequestBody requestBody = RequestBody.create(body, MediaType.parse("application/json"));

    Request request;
    try {
      request = new Request.Builder().post(requestBody).url(uri.toURL()).headers(headers()).build();
    } catch (TokenException e) {
      throw new IOException(e);
    }
    try (Response response = httpClient.newCall(request).execute()) {
      return handleResponse(response);
    }
  }

  public Map<String, Object> getUserDetails(String projectId, String userId) throws IOException {
    URI uri = URI.create(baseUrl).resolve("/projects/" + projectId + "/users/" + userId);

    Request request;
    try {
      request = new Request.Builder().get().url(uri.toURL()).headers(headers()).build();
    } catch (TokenException e) {
      throw new IOException(e);
    }

    try (Response response = httpClient.newCall(request).execute()) {
      return handleResponse(response);
    }
  }

  private Map<String, Object> handleResponse(Response response) throws IOException {

    switch (response.code()) {
      case 404:
        throw new IOException("The Entity or URL was not found in the appserver." + response);
      case 200:
      case 201:
        if (response.body() == null) {
          throw new IOException(
              "The response was success from appserver but body was null. " + response);
        }
        return objectMapper.readValue(response.body().string(), new TypeReference<>() {});
      default:
        throw new IOException("There was an error requesting the appserver. " + response);
    }
  }
}

File: src/main/java/org/radarbase/appserver/client/AppserverClient.kt
package org.radarbase.appserver.client

import com.fasterxml.jackson.core.type.TypeReference
import com.fasterxml.jackson.databind.ObjectMapper
import okhttp3.*
import okhttp3.Headers.Companion.headersOf
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.RequestBody.Companion.toRequestBody
import org.radarbase.exception.TokenException
import org.radarbase.oauth.OAuth2Client
import org.slf4j.LoggerFactory
import java.io.IOException

/**
 * Client to provide interface to the AppServer REST API. It uses the OAuth client library from
 * management portal to authenticate requests.
 */
class AppserverClient(config: AppserverClientConfig) {
    private val baseUrl: HttpUrl
    private val httpClient: OkHttpClient = config.httpClient ?: OkHttpClient()
    private val objectMapper: ObjectMapper = config.mapper ?: ObjectMapper()
    private val oauthClient: OAuth2Client?

    constructor(build: AppserverClientConfig.() -> Unit): this(AppserverClientConfig().apply(build))

    init {
        baseUrl = requireNotNull(config.appserverUrl) { "Appserver client needs base URL" }

        oauthClient = if (config.tokenUrl == null) {
            logger.warn("MP Token url was not provided. Will use unauthenticated requests to appserver.")
            null
        } else {
            val clientId = requireNotNull(config.clientId) { "Client ID for appserver client is not set." }
            val clientSecret = requireNotNull(config.clientSecret) { "Client secret for appserver client is not set." }
            OAuth2Client.Builder()
                .credentials(clientId, clientSecret)
                .endpoint(config.tokenUrl)
                .scopes("SUBJECT.READ SUBJECT.UPDATE PROJECT.READ")
                .httpClient(httpClient)
                .build()
        }
    }

    @Throws(TokenException::class)
    private fun headers(): Headers {
        return if (oauthClient != null) {
            headersOf("Authorization", "Bearer " + oauthClient.validToken.accessToken)
        } else {
            headersOf()
        }
    }

    @Throws(IOException::class)
    fun createMessage(
        projectId: String,
        userId: String,
        type: MessagingType,
        contents: AppServerMessageContents,
    ): Map<String, Any> {
        val stringContents = objectMapper.writeValueAsString(contents)
        return createMessage(projectId, userId, type, stringContents)
    }

    @Throws(IOException::class)
    fun createMessage(
        projectId: String,
        userId: String,
        type: MessagingType,
        contents: String,
    ): Map<String, Any> {
        val request: Request = try {
            Request.Builder()
                .post(contents.toRequestBody(APPLICATION_JSON))
                .url(
                    baseUrl.newBuilder()
                        .addEncodedPathSegment("projects")
                        .addPathSegment(projectId)
                        .addEncodedPathSegment("users")
                        .addPathSegment(userId)
                        .addEncodedPathSegment("messaging")
                        .addEncodedPathSegment(type.urlPart)
                        .build()
                )
                .headers(headers())
                .build()
        } catch (e: TokenException) {
            throw IOException(e)
        }
        return httpClient.newCall(request).execute().use { handleResponse(it) }
    }

    @Throws(IOException::class)
    fun getUserDetails(projectId: String, userId: String): Map<String, Any> {
        val request: Request = try {
            Request.Builder()
                .get()
                .url(
                    baseUrl.newBuilder()
                        .addEncodedPathSegment("projects")
                        .addPathSegment(projectId)
                        .addEncodedPathSegment("users")
                        .addPathSegment(userId)
                        .build()
                )
                .headers(headers())
                .build()
        } catch (e: TokenException) {
            throw IOException(e)
        }
        return httpClient.newCall(request).execute().use { handleResponse(it) }
    }

    @Throws(IOException::class)
    private fun handleResponse(response: Response): Map<String, Any> {
        val body = response.body
            ?: throw IOException("Response body from appserver was null.")

        return when (response.code) {
            404 -> throw IOException("The Entity or URL was not found in the appserver: $body")
            200, 201 -> objectMapper.readValue(body.string(), object : TypeReference<Map<String, Any>>() {})
            else -> throw IOException("There was an error requesting the appserver. $response")
        }
    }

    companion object {
        private val logger = LoggerFactory.getLogger(AppserverClient::class.java)

        private val APPLICATION_JSON = "application/json".toMediaType()
    }
}
